app/src/lib/find-default-branch.ts

```diff
@@ -0,0 +1,56 @@
+import { Branch, BranchType } from '../models/branch'
+import { Repository } from '../models/repository'
+import { getRemoteHEAD } from './git'
+import { getDefaultBranch } from './helpers/default-branch'
+
+/**
+ * Attempts to locate the default branch as determined by the HEAD symbolic link
+ * in the contribution target remote (origin or upstream) if such a ref exists,
+ * falling back to the value of the `init.defaultBranch` configuration and
+ * finally a const value of `main`.
+ *
+ * In determining the default branch we prioritize finding a local branch but if
+ * no local branch matches the default branch name nor is tracking the
+ * contribution target remote HEAD we'll fall back to looking for the remote
+ * branch itself.
+ */
+export async function findDefaultBranch(
+  repository: Repository,
+  branches: ReadonlyArray<Branch>
+) {
+  const originHead = await getRemoteHEAD(repository, 'origin')
+  const defaultBranchName = originHead ?? (await getDefaultBranch())
+  const remoteRef = originHead ? `origin/${originHead}` : undefined
+
+  let localHit: Branch | undefined = undefined
+  let localTrackingHit: Branch | undefined = undefined
+  let remoteHit: Branch | undefined = undefined
+
+  for (const branch of branches) {
+    if (branch.type === BranchType.Local) {
+      if (branch.name === defaultBranchName) {
+        localHit = branch
+      }
+
+      if (remoteRef && branch.upstream === remoteRef) {
+        // Give preference to local branches that target the upstream
+        // default branch that also match the name. In other words, if there
+        // are two local branches which both track the origin default branch
+        // we'll prefer a branch which is also named the same as the default
+        // branch name.
+        if (!localTrackingHit || branch.name === defaultBranchName) {
+          localTrackingHit = branch
+        }
```

Comment:
Nit - not likely we have many branches. But, since we are giving preference to the `localTrackingHit`, seems we can short circuit the `for..loop` once it is found.

```suggestion
        if (!localTrackingHit || branch.name === defaultBranchName) {
          localTrackingHit = branch
          break
        }
```

Reply:
Originedeffault

Reply:
I don't think we can because `localTrackingHit` can be set more than once. It'll be set as soon as we find a branch that's tracking the remote HEAD but if there are more than one local branch tracking the remote HEAD and we see one where the local branch name that also matches the default branch name we'll re-assign it to that.